### ISSUE TITLE
Dilated Convolution, Dilated Pooling and 2D/3D Sliding Window CNNs

### ISSUE THREAD

**Issue by psicalculus** [tensorflow/tensorflow#3492]
Could you please implement 2D/3D dilated convolution and 2D/3D dilated pooling in tensorflow?

Please see http://arxiv.org/pdf/1511.07122.pdf for a reference on dilated convolution.

Dilated max-pooling is simply regular max-pooling but the pixels/voxels you use in each "application" of the max-pooling operation are exactly the same pixels/voxels you would select with dilated convolution.

Dilated convolution/pooling are useful for connectomics and 3D shape datasets (3D deep learning).

**Comment by psicalculus** [comment]
It looks like tensorflow already supports 2D dilated convolution.

``tf.nn.atrous_conv2d(value, filters, rate, padding, name=None)``

**Comment by gpapan** [comment]
(1) We do not have immediate plans to add support for 3D atrous/dilated convolution in TF. For the 2-D case we already have `tf.nn.atrous_conv2d(value, filters, rate, padding, name=None)`.

(2) You can adapt the existing [tf.nn.atrous_conv2d] (https://github.com/tensorflow/tensorflow/blob/73ced9d797056c7e67a06ed2098dd809d85ec44a/tensorflow/python/ops/nn_ops.py) implementation to perform 2D atrous/dilated pooling by replacing the call to `conv2d` with a call to `avg_pool` or `max_pool`.

An alternative to achieve a similar effect (i.e., pooling over a larger area) is to just use standard pooling with enlarged kernel size:

     ksize_eff = ksize + (ksize - 1) * (rate - 1)

Note that average or max pooling act on each channel separately, which means that they are fast even for large kernel size. They also have no learned parameters, which means that there is no risk for over-fitting. These factors combined imply that kernel upsampling is less appealing for pooling than it is for convolution.

**Comment by psicalculus** [comment]
On connectomics datasets, the 3D images included are very high resolution. 3D neural nets trained on such datasets have lots and lots of parameters. For this reason, 3D dilated convolution and 3D dilated pooling are very useful in speeding up the learning process and enabling the training of very deep nets (cover larger area but less computation/less parameters to train). Most deep learning frameworks including TensorFlow are optimized to run 2D deep nets fast, not 3D deep nets.

**Comment by psicalculus** [comment]
In addition, support for 2D/3D sliding window CNNs (specifically, dense training using output patches) would be appreciated.

For references on 2D/3D sliding window CNNs, see:
- http://znn-release.readthedocs.io/en/latest/
- http://people.idsia.ch/~juergen/nips2012.pdf

**Comment by michaelisard** [comment]
(Edited title and added 'Enhancement' label per convention.)

**Comment by michaelisard** [comment]
Marking 'contributions welcome' since this would be great to add, but isn't currently being worked on.

**Comment by KendallWeihe** [comment]
Any word on the progress of this? I'm looking for 3D support

**Comment by jbms** [comment]
This is being implemented --- hopefully will be available soon.

**Comment by hmartiro** [comment]
+1 for 3D dilated convolutions.

Torch implements this with nn.VolumetricDilatedConvolution:
https://github.com/torch/nn/blob/master/doc/convolution.md#nn.VolumetricDilatedConvolution

**Comment by jbms** [comment]
This is implemented in master --- see tf.nn.convolution and tf.nn.pool.

**Comment by ankitvgupta** [comment]
Is there any timeline for implementing tf.nn.pool with "SAME" padding with dilation_rate > 1? It looks like tf.nn.atrous_conv2d works with SAME padding, but the tf.nn.pool function doesn't have it.

**Comment by KendallWeihe** [comment]
Any word on if dilated convolutions have been implemented in the convolutional transpose layers? Or a timeline on when it can be?

**Comment by ahundt** [comment]
I believe this is implemented in `tensorflow/models/slim/.../resnet_utils.py` in the function [conv2d_same](https://github.com/tensorflow/models/blob/master/slim/nets/resnet_utils.py#L77). I explained in more detail at https://github.com/tensorflow/tensorflow/issues/4742#issuecomment-285931111.

**Comment by girving** [comment]
Looks like a good chunk of this is done.  Further requests should probably be filed as further issues.
